### PR TITLE
Reduce string concat in hibernate instrumentation

### PR DESCRIPTION
**Pull request by tylerbenson**

### CODE REVIEW

File: dd-java-agent/instrumentation/hibernate/src/main/java/datadog/trace/instrumentation/hibernate/SessionMethodUtils.java [hunk 1]
```diff
@@ -24,6 +25,7 @@
   public static <TARGET, ENTITY> SessionState startScopeFrom(
       final ContextStore<TARGET, SessionState> contextStore,
       final TARGET spanKey,
+      final Method origin,
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/hibernate/src/main/java/datadog/trace/instrumentation/hibernate/SessionMethodUtils.java:28]: Any particular reason to pass along this, even though it's not used?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java [hunk 2]
```diff
@@ -1,5 +1,8 @@
 package datadog.trace.bootstrap.instrumentation.api;
 
+import lombok.EqualsAndHashCode;
+
+@EqualsAndHashCode
```
Comment (bantonsson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java:5]: This seems unrelated to the change.

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java [hunk 3]
```diff
@@ -44,13 +45,15 @@
 
     @Advice.OnMethodEnter(suppress = Throwable.class)
     public static SessionState startMethod(
-        @Advice.This final Criteria criteria, @Advice.Origin("#m") final String name) {
+        @Advice.This final Criteria criteria,
+        @Advice.Origin("hibernate.criteria.#m") final String operationName,
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: Ah, nice. I assume that this is _cached_ at the call site.

Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: Is 16 a good guess for the size of the hash map? That would imply an expected cardinality of `operationName` of about 8. It might be reasonable, but I'd want to know why.

File: dd-java-agent/instrumentation/hibernate/src/main/java/datadog/trace/instrumentation/hibernate/SessionMethodUtils.java (hunk 1 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/hibernate/src/main/java/datadog/trace/instrumentation/hibernate/SessionMethodUtils.java:28]: This is related to another PR I'm working on... which inspired this change.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java:5]: yeah, unrelated now.  I originally was caching with a pair as the key, but then I figured out the call site caching option and decided to do that instead.  I left this in here because it seemed like an oversight that might be missed later.  Would you prefer I pull it out?

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: I'm assuming that this avoids repeat concatenation by creating a constant in the byte code.  I didn't actually verify that though.

Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: But the question was why the guess of the size of 16? What's the rationale for choosing 16?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java (hunk 2 above)
Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Pair.java:5]: Not sure it's an oversight, do we have any `Pair`s as keys in maps where this would be relevant? Not opposed to adding it though.

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: Ok, I ran it in a debugger and verified that the object instance is reused, where on master it creates a new one each time.

Comment (richardstartin) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java:49]: OK. But why 16? What's the reasoning? How do we know this cache won't just thrash? We can't get it all the time with a fixed size but I'm just asking what your reasoning is.
